src/Sylius/Component/Core/OrderProcessing/ShippingChargesProcessor.php

```diff
@@ -47,16 +47,24 @@ public function process(BaseOrderInterface $order): void
         $order->removeAdjustments(AdjustmentInterface::SHIPPING_ADJUSTMENT);
 
         foreach ($order->getShipments() as $shipment) {
+            $shipment->removeAdjustments(AdjustmentInterface::SHIPPING_ADJUSTMENT);
+
             try {
                 $shippingCharge = $this->shippingChargesCalculator->calculate($shipment);
+                $shippingMethod = $shipment->getMethod();
 
                 /** @var AdjustmentInterface $adjustment */
                 $adjustment = $this->adjustmentFactory->createNew();
                 $adjustment->setType(AdjustmentInterface::SHIPPING_ADJUSTMENT);
                 $adjustment->setAmount($shippingCharge);
-                $adjustment->setLabel($shipment->getMethod()->getName());
+                $adjustment->setLabel($shippingMethod->getName());
+                $adjustment->setDetails([
+                    'shippingMethodCode' => $shippingMethod->getCode(),
+                    'shippingMethodName' => $shippingMethod->getName(),
+                ]);
```

Comment:
Just a thought for improvement. What do you think about some factory service, that would be responsible for this array creation? Seems to provide a more easier way to customize its structure.

Reply:
I agree with the need to freeze data format for details, but external service to handle internal state of model is a bit of anti-pattern. (Something along the lines of anemic domain model)

Model should be able to handle format on it's own, for example like so
```php
$adjustment->setShippingMethodDetails($shippingMethod->getCode(), $shippingMethod->getName());
```

```php
// Adjustment

public function setShippingMethodDetails(string $shippingMethodCode, string $shippingMethodName): void
{
    $this->details['shippingMethodCode'] = $shippingMethodCode;
    $this->details['shippingMethodName'] = $shippingMethodName;
}
```

or alternatively as set of value objects.